resdk/query.py

```diff
@@ -0,0 +1,246 @@
+            * slug (=)
+            * descriptor (=)
+            * data (=)
+            * descriptor_schema (=)
+            * id (=)
+
+    Example usage:
+    # Get a list of data objects with status set to OK.
+    re.data.filter(status='OK')
+    # Get a liust of sample objects that contain data object 42 and
+    # were contributed by contibutor with ID 1
+    re.collection.filter(data=42, contributor=1)
+
+    """
+
+    _cache = None
+    _limit = None
+    _offset = None
+    _filters = collections.defaultdict(list)
+
+    resolwe = None
+    resource = None
+    endpoint = None
+    api = None
+    logger = None
+
+    def __init__(self, resolwe, Resource, endpoint=None):
+        """Initialize attributes."""
+        self.resolwe = resolwe
+        self.resource = Resource
+        self.endpoint = endpoint if endpoint else Resource.endpoint
+        self.api = operator.attrgetter(self.endpoint)(resolwe.api)
+        self.logger = logging.getLogger(__name__)
+
+    def __getitem__(self, index):
+        """Retrieve an item or slice from the set of results."""
+        # pylint: disable=protected-access
+        if not isinstance(index, (slice,) + six.integer_types):
+            raise TypeError
+        if ((not isinstance(index, slice) and index < 0) or
+                (isinstance(index, slice) and index.start is not None and index.start < 0) or
+                (isinstance(index, slice) and index.stop is not None and index.stop < 0)):
+            raise ValueError("Negative indexing is not supported.")
+        if isinstance(index, slice) and index.step is not None:
+            raise ValueError("`step` parameter in slice is not supported")
+
+        if self._cache is not None:
+            return self._cache[index]
+
+        new_query = self._clone()
+
+        if isinstance(index, slice):
+            if self._offset or self._limit:
+                raise NotImplementedError('You cannot slice already sliced query.')
+
+            start = 0 if index.start is None else int(index.start)
+            stop = 1000000 if index.stop is None else int(index.stop)  # default to something big
+            new_query._offset = start
+            new_query._limit = stop - start
+            return new_query
+
+        new_query._offset = self._offset + index if self._offset else index
+        new_query._limit = 1
+
+        query_list = list(new_query)
+        if len(query_list) == 0:
+            raise IndexError('list index out of range')
+        return query_list[0]
+
+    def __iter__(self):
+        """Return itterator over current the object."""
+        self._fetch()
+        return iter(self._cache)
+
+    def __repr__(self):
+        """Return string representation of the current object."""
+        self._fetch()
+        return '[{}]'.format(',\n '.join(str(obj) for obj in self._cache))
+
+    def __len__(self):
+        """Return length of results og current query."""
+        self._fetch()
+        return len(self._cache)
```

Comment:
In not cached, could we just query for the number of objects on the server instead of transferring down all objects?